File: NotJSBEditor/Rendering/MeshRenderer.cs
using OpenTK.Graphics.OpenGL4;
using System;

namespace NotJSBEditor.Rendering
{
    public class MeshRenderer : IRenderer
    {
        public Mesh Mesh;
        public Material Material;

        private int transformBuffer;

        private int[] uniformBuffers;

        // Initialize resources
        public MeshRenderer()
        {
            // Make a transform UBO
            transformBuffer = GL.GenBuffer();
            GL.BindBuffer(BufferTarget.UniformBuffer, transformBuffer);
            GL.BufferData(BufferTarget.UniformBuffer, 256, IntPtr.Zero, BufferUsageHint.DynamicCopy);
            GL.BindBuffer(BufferTarget.UniformBuffer, 0);

            // Initialize buffer array
            uniformBuffers = new int[]
            {
                transformBuffer
            };
        }

        public bool Render(InputDrawData drawData, out OutputDrawData outDrawData)
        {
            outDrawData = new OutputDrawData();

            if (!CanRender())
                return false;

            // Construct transform buffer struct
            TransformBuffer transform = new TransformBuffer()
            {
                Model = drawData.Model,
                View = drawData.View,
                Projection = drawData.Projection,
                ModelViewProjection = drawData.ModelViewProjection
            };

            // Update UBO
            UpdateTransformBufer(transform);

            // Construct the draw data for rendering
            outDrawData = new OutputDrawData()
            {
                VAO = Mesh.VAO,
                Shader = Material.Shader.Handle,
                UniformBuffers = uniformBuffers,
                Command = new DrawElementsCommand()
                {
                    Count = Mesh.IndicesCount,
                    Offset = 0
                }
            };

            return true;
        }

        private void UpdateTransformBufer(TransformBuffer transform)
        {
            GL.BindBuffer(BufferTarget.UniformBuffer, transformBuffer);
            GL.BufferSubData(BufferTarget.UniformBuffer, IntPtr.Zero, 256, ref transform);
            GL.BindBuffer(BufferTarget.UniformBuffer, 0);
        }

        // Check if rendering is possible
        private bool CanRender()
        {
            if (Mesh == null || Material == null)
                return false;

            return true;
        }

        // Free unmanaged resources
        public void Dispose()
        {
            GL.DeleteBuffer(transformBuffer);
        }
    }
}

File: NotJSBEditor/src/rendering/MeshRenderer.cpp
#include "MeshRenderer.h"
MeshRenderer::MeshRenderer()
{
	mesh = nullptr;
	material = nullptr;
}

MeshRenderer::~MeshRenderer()
{
}

bool MeshRenderer::render(InputDrawData input, OutputDrawData** output)
{
	if (!canRender())
	{
		return false;
	}

	if (opacity == 0.0f)
	{
		return false;
	}

	// Decompose matrix
	glm::vec3 translation = input.model[3];

	// Construct the draw data for rendering
	OutputDrawData* drawData = new OutputDrawData();
	drawData->mesh = mesh;
	drawData->material = material;
	drawData->transform = input.model;
	drawData->drawDepth = translation.z;
	drawData->drawTransparent = (opacity < 1.0f);
	drawData->opacity = opacity;

	*output = drawData;

	return true;
}

bool MeshRenderer::canRender() const
{
	if (mesh == nullptr || material == nullptr)
		return false;

	return true;
}
